### DISCUSSION TITLE
Understanding scope and its interaction with nn.Module

### DISCUSSION THREAD

**Discussion by Jeevesh8** [google/flax#2368] [Q&A] [answered]
Is there a design document explaining how exactly the ``core.scope`` allows parameters to get injected through the copy of ``nn.Module`` created in apply? In particular, how exactly the variables are assigned to the nested submodules. 

I am trying to understand this from the code-base, but it is taking too long. Does anyone know if there exists any tutorial or such. If not maybe I will write one, afterwards..

**Comment by levskaya** [comment, marked as answer]
You should just think of each `nn.Module` `self` carrying a single `scope`, there's always a 1:1 relationship between a Module instance and a Scope object - most of the scope api like `param`, `variable`, `put/get/has_variable` are more or less passed through.  Modules just help deal with sharing by reference, handling multiple methods or setup methods, etc.

The modules are populated with scopes when the top-level module is used with a `apply`, `init`, or `bind` method.  Details at:
https://flax.readthedocs.io/en/latest/advanced_topics/module_lifecycle.html
